src/spurt/io/_three_d.py

```diff
@@ -0,0 +1,263 @@
+            raise ValueError(errmsg)
+
+        # We only hold a reference to the data
+        self._data = data
+        self._time_dim = time_dim
+
+    @property
+    def dtype(self) -> DTypeLike:
+        return self._data.dtype
+
+    @property
+    def ndim(self) -> int:
+        return 2
+
+    @property
+    def time_dim(self) -> int:
+        return int(self._time_dim > 0)
+
+    @property
+    def space_dim(self) -> int:
+        return int(self._time_dim == 0)
+
+    @property
+    def _space_shape(self) -> tuple[int, ...]:
+        return tuple(
+            ss for ii, ss in enumerate(self._data.shape) if ii != self._time_dim
+        )
+
+    @property
+    def _space_size(self) -> int:
+        return int(np.prod(self._space_shape))
+
+    @property
+    def _time_size(self) -> int:
+        return self._data.shape[self._time_dim]
+
+    @property
+    def shape(self) -> tuple[int, ...]:
+        if self.time_dim:
+            return (self._space_size, self._time_size)
+
+        return (self._time_size, self._space_size)
+
+    @property
+    def xcoords(self) -> ArrayLike:
+        """Return the x-coordinate of the points as an array.
+
+        This is to mimic an irregular grid.
+        """
+        return np.arange(np.prod(self._data.shape), dtype=int) % self._space_shape[1]
+
+    @property
+    def ycoords(self) -> ArrayLike:
+        """Return the y-coordinate of the points as an array.
+
+        This is to mimic an irregular grid.
+        """
+        return np.arange(np.prod(self._data.shape), dtype=int) // self._space_shape[1]
+
+    def get_time_slice(self, key: int) -> ArrayLike:
+        """Read a block of data in time.
+
+        Only makes a copy if absolutely needed.
+        """
+        ind = tuple(
+            slice(None) if ii != self._time_dim else int(key)
+            for ii in range(self._data.ndim)
+        )
+        return self._data[ind].ravel()  # type: ignore[arg-type, index]
+
+    def get_spatial_slice(self, key: int) -> ArrayLike:
+        """Read a block of data in space.
+
+        Only makes a copy if absolutely needed.
+        """
+        ind = list(np.unravel_index(key, self._space_shape))
+        ind.insert(self._time_dim, slice(None))
+
+        return self._data[tuple(ind)].ravel()  # type: ignore[arg-type, index]
+
+
+class Irreg3DInput(InputStackInterface):
+    """A single numpy 2D array as input.
+
+    One axis represents time and the other some flattened representation
+    in space. `time_dim` is an optional input to indicate time axis.
+    """
+
+    def __init__(
+        self,
+        data: InputInterface,
+        xy: InputInterface,
+        time_dim: int = 0,
+    ):
+        """Initialize a stack interface for a 2D numpy array.
+
+        One axis represents time axis and the other spatial axis.
+
+        Parameters
+        ----------
+        data: np.ndarray
+            Numpy 2D array as input
+
+        xy: np.ndarray
+            Numpy npts x 2 array as input
```

Comment:
I'm a little unclear about what `xy` means.

Are these the (row,col) indices of each point from the original image array? If so, could you modify the description to something like this?

```suggestion
            An Nx2 array of (row,col) indices of each pixel in space,
            where N is the number of spatial points.
```

(Maybe I got that backwards and `x` are supposed to be the col indices `y` are supposed to be the row indices.)

Or are `x` and `y` supposed to be real-space coordinates, like e.g. longitude & latitude?

Reply:
It is generic enough that it can be either - and will depend on the context in which the object gets used.